server/src/main/java/com/cloud/network/IpAddressManagerImpl.java

```diff
@@ -1476,7 +1476,8 @@ public IPAddressVO associateIPToGuestNetwork(long ipId, long networkId, boolean
         //     - if shared network in Advanced zone
         //     - and it belongs to the system
         if (network.getAccountId() != owner.getId()) {
-            if (zone.getNetworkType() != NetworkType.Basic && !(zone.getNetworkType() == NetworkType.Advanced && network.getGuestType() == Network.GuestType.Shared)) {
+            if (zone.getNetworkType() != NetworkType.Basic &&
+                    !(zone.getNetworkType() == NetworkType.Advanced && network.getGuestType() == Network.GuestType.Shared || network.getVpcId().equals(ipToAssoc.getVpcId())))  {
```

Comment:
So if `network.getVpcId().equals(ipToAssoc.getVpcId())` we should throw an exception! Can you explain?

Reply:
Theres a subtle "!" in the beginning of the expression
`(zone.getNetworkType() == NetworkType.Advanced && network.getGuestType() == Network.GuestType.Shared || network.getVpcId().equals(ipToAssoc.getVpcId()))`
which makes it equivalent to the following expression
`(!zone.getNetworkType() == NetworkType.Advanced && !network.getVpcId().equals(ipToAssoc.getVpcId()) || !network.getGuestType() == Network.GuestType.Shared && !network.getVpcId().equals(ipToAssoc.getVpcId()))`

So we are actually checking if both VpcIds are different and throwing an exception if that is the case.

Reply:
right, missed that. Maybe you can extract the expression to a method for clarity?
thanks for the explanation.